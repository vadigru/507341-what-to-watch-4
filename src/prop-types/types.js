import PropTypes from "prop-types";

export const movieType = PropTypes.shape({
  title: PropTypes.string.isRequired,
  posterUrl: PropTypes.string.isRequired,
  backgroundUrl: PropTypes.string.isRequired,
  // backgroundColor: PropTypes.string.isRequired,
  previewUrl: PropTypes.string.isRequired,
  previewImage: PropTypes.string.isRequired,
  genre: PropTypes.string.isRequired,
  release: PropTypes.number.isRequired,
  director: PropTypes.string.isRequired,
  starring: PropTypes.arrayOf(PropTypes.string.isRequired),
  time: PropTypes.string.isRequired,
  rating: PropTypes.number.isRequired,
  votes: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  // isFavorite: PropTypes.bool.isRequired,
  videoUrl: PropTypes.string.isRequired,
});
