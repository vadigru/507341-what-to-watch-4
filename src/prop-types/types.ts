export type Movie = {
  title: string,
  posterUrl: string,
  backgroundUrl: string,
  backgroundColor: string,
  previewUrl: string,
  previewImage: string,
  genre: string,
  release: number,
  director: string,
  starring: string[],
  time: string,
  rating: number,
  votes: number,
  description: string,
  id: number,
  isFavorite: boolean,
  videoUrl: string,
};
