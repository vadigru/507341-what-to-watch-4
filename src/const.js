export const MOVIES_DEFAULT_AMOUNT = 8;
export const MOVIES_STEP_AMOUNT = 8;
export const MAX_GENRES_AMOUNT = 9;
export const ALL_GENRES = `All genres`;
export const START_DELAY = 1000;

export const Tab = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const DefaultTab = {
  GENRE: `All genres`,
  MOVIE: `Overview`
};

export const RatingLevel = {
  ZERO: 0,
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  AWESOME: 10
};

export const AppRoute = {
  ROOT: `/`,
  MOVIE_PAGE: `/movie`,
  SIGN_IN: `/login`,
  ADD_REVIEW: `/review`,
};
