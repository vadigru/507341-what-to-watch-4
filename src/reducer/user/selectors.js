import Namespace from "../namespace.js";

export const getAuthorizationStatus = (state) => {
  return state[Namespace.USER].authorizationStatus;
};

export const getInvalidAuthorizationStatus = (state) => {
  return state[Namespace.USER].isValidAuthorization;
};

export const getAvatar = (state) => {
  return state[Namespace.USER].avatarUrl;
};
