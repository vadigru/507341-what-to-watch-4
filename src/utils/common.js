import {RatingLevel, ALL_GENRES} from "../const.js";

export const getTextedRating = (rating) => {
  switch (true) {
    case rating >= RatingLevel.ZERO && rating <= RatingLevel.BAD:
      return `Bad`;
    case rating > RatingLevel.BAD && rating <= RatingLevel.NORMAL:
      return `Normal`;
    case rating > RatingLevel.NORMAL && rating <= RatingLevel.GOOD:
      return `Good`;
    case rating > RatingLevel.GOOD && rating < RatingLevel.AWESOME:
      return `Very good`;
    default:
      return `Awesome`;
  }
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const getMoviesByGenre = (genre, movies) => {
  return genre === ALL_GENRES
    ? movies
    : movies.filter((movie) => movie.genre === genre);
};
