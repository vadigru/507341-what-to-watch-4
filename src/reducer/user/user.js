import {extend} from "../../utils/common.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`
};

const SERVER_URL = `https://4.react.pages.academy`;

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  isValidAuthorization: true,
  avatarUrl: ``,
};

const ActionType = {
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  IS_INVALID_AUTHORIZATION: `IS_INVALID_AUTHORIZATION`,
  SET_AVATAR_URL: `SET_AVATAR_URL`
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRE_AUTHORIZATION,
    payload: status
  }),
  requireValidAuthorization: (isValid) => ({
    type: ActionType.IS_INVALID_AUTHORIZATION,
    payload: isValid
  }),
  setAvatarUrl: (avatarUrl) => ({
    type: ActionType.SET_AVATAR_URL,
    payload: avatarUrl
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRE_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload
      });
    case ActionType.IS_INVALID_AUTHORIZATION:
      return extend(state, {
        isValidAuthorization: action.payload
      });
    case ActionType.SET_AVATAR_URL:
      return extend(state, {
        avatarUrl: `${SERVER_URL}${action.payload}`
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
    .then((response) => {
      const {avatar_url: avatar} = response.data;
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAvatarUrl(avatar));
    })
    .catch((err) => {
      throw err;
    });
  },

  login: (authData) => (dispatch, getState, api) => {
    // dispatch(ActionCreator.requireValidAuthorization(true));
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password
    })
    .then((response) => {
      const {avatar_url: avatar} = response.data;
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAvatarUrl(avatar));
    })
    .catch((err)=>{
      dispatch(ActionCreator.requireValidAuthorization(false));
      throw err;
    });
  }
};

export {ActionCreator, ActionType, AuthorizationStatus, Operation, reducer};
