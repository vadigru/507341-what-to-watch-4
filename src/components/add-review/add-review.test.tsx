import * as React from "react";
import * as renderer from "react-test-renderer";
import AddReview from "./add-review";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {ALL_GENRES, MOVIES_DEFAULT_AMOUNT} from "../../const";
import Namespace from "../../reducer/namespace";
import {AuthorizationStatus} from "../../reducer/user/user";
import {Router} from "react-router-dom";
import history from "../../history";
import {Movie} from "../../prop-types/types";
import {noop} from "../../utils/common";

const mockStore = configureStore([]);

const movie: Movie = {
  title: `Movie Name`,
  posterUrl: `https://url.com`,
  backgroundUrl: `https://url.com`,
  backgroundColor: `some color`,
  previewUrl: `https://url.com`,
  previewImage: `https://url.com`,
  genre: `genre`,
  release: 2020,
  director: `Famous Director`,
  starring: [`Actor One`, `Actor Two`, `Actor Three`],
  time: `1h 30m`,
  rating: 10,
  votes: 1000000,
  description: `Some Description`,
  id: 1,
  isFavorite: true,
  videoUrl: `https://url.com`,
};

it(`Should render AddReview component`, () => {
  const store = mockStore({
    [Namespace.DATA]: {
      films: [],
      promo: {},
      reviews: [],
      isFilmsLoading: false,
      isPromoLoading: false,
      isReviewsLoading: false,
      isReviewPosting: false,
      isReviewSendingError: false,
    },
    [Namespace.STATE]: {
      genre: ALL_GENRES,
      showedMovies: MOVIES_DEFAULT_AMOUNT,
      selectedMovieId: 1
    },
    [Namespace.USER]: {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
      isValidAuthorization: true,
      avatarUrl: ``,
    },
  });

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <AddReview
              id={1}
              movie={movie}
              isReviewPosting={false}
              isReviewSendingError={false}
              avatarUrl={``}
              rating={0}
              comment={0}
              onFormDataChange={noop}
            />
          </Router>
        </Provider>
    )
  .toJSON();

  expect(tree).toMatchSnapshot();
});
