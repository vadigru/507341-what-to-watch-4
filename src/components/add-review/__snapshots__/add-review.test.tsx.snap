// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddReview component 1`] = `"ERROR"`;
